examples/finetune_resnet/components.py

```diff
@@ -0,0 +1,138 @@
+# reformats the data so that it is suitable for the model.
+@DataCollator.register("image_collator")
+class ImageCollator(DataCollator[Dict[str, Any]]):
+    def __call__(self, batch: List[Dict[str, Any]]) -> Dict[str, Any]:
+        return {
+            "image": torch.cat([item["image"].unsqueeze(0) for item in batch], dim=0),
+            "label": torch.tensor([item["labels"] for item in batch]),
+        }
+
+
+# Function that returns an image transformations dict with the appropriate image size.
+def get_data_transforms(input_size: int):
+    data_transforms = {
+        "train": transforms.Compose(
+            [
+                transforms.RandomResizedCrop(input_size),
+                transforms.RandomHorizontalFlip(),
+                transforms.ToTensor(),
+                transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
+            ]
+        ),
+        "test": transforms.Compose(
+            [
+                transforms.Resize(input_size),
+                transforms.CenterCrop(input_size),
+                transforms.ToTensor(),
+                transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
+            ]
+        ),
+    }
+    return data_transforms
+
+
+# loads and image and applies the appropriate transformation
+def pil_loader(path: str, input_size: int, transform_type: str):
+    with open(path, "rb") as f:
+        image = Image.open(f)
+        image = image.convert("RGB")
+        transform = get_data_transforms(input_size=input_size)[transform_type]
+        transformed_image = transform(image)
+        return transformed_image
+
+
+# calls the image loader on every image in a given batch
+def image_loader(example_batch, input_size: int, transform_type: str):
+    example_batch["image"] = [
+        pil_loader(f, input_size, transform_type) for f in example_batch["file"]
+    ]
+    return example_batch
+
+
+# This step takes in raw image data and transforms and tokenizes it.
+@Step.register("transform_data")
+class TransformData(Step):
+    DETERMINISTIC = True
+    CACHEABLE = False
+
+    def run(  # type: ignore
+        self, dataset: datasets.DatasetDict, test_size: float, input_size: int
+    ) -> datasets.DatasetDict:
+        def image_loader_wrapper(example_batch):
+            return image_loader(example_batch, input_size=input_size, transform_type="train")
+
+        dataset = dataset.with_transform(image_loader_wrapper)
+        train_test = dataset["train"].train_test_split(test_size=test_size)
+        return train_test
+
+
+# function to map integer labels to string labels
+def convert_to_label(int_label: int) -> str:
+    if int_label == 0:
+        return "cat"
+    else:
+        return "dog"
+
+
+@Step.register("prediction")
+class Prediction(Step):
+    FORMAT: Format = JsonFormat()
+
+    def run(self, image_url: str, input_size: int, model: models) -> Dict[str, Any]:  # type: ignore
```

Comment:
This should also take a device parameter.
```suggestion
    def run(self, image_url: str, input_size: int, model: models, device: Optional[str] = "cpu") -> Dict[str, Any]:  # type: ignore
        device = torch.device(device)
        model = model.to(device)
```

Reply:
added